crates/sui-core/src/checkpoints/causal_order.rs

```diff
@@ -0,0 +1,397 @@
+            .extra_transactions
+            .keys()
+            .map(|e| e.transaction)
+            .collect();
+
+        // Index the effects by transaction digest, as we will need to look them up.
+        let mut effect_map: BTreeMap<TransactionDigest, &TransactionEffects> = effects
+            .iter()
+            .map(|e| {
+                let e = e.as_ref().unwrap();
+                (e.transaction_digest, e)
+            })
+            .collect();
+
+        // Only include in the checkpoint transactions that have not been checkpointed before.
+        // Due to the construction of the checkpoint table `extra_transactions` and given that
+        // we must have processed all transactions in the proposed checkpoint, this check is
+        // reduced to ensuring that the transactions are in the table `extra_transactions`
+        // (that lists transactions executed but not yet checkpointed).
+        let digest_map: BTreeMap<TransactionDigest, &ExecutionDigests> = digests
+            .iter()
+            .filter_map(|d| {
+                if tx_not_in_checkpoint.contains(&d.transaction) {
+                    Some((d.transaction, d))
+                } else {
+                    // We remove the effects map entries for transactions
+                    // that are already checkpointed.
+                    effect_map.remove(&d.transaction);
+                    None
+                }
+            })
+            .collect();
+
+        // Build a forward index of transactions. This will allow us to start with the initial
+        // and then sequenced trasnactions and efficiently determine which other transactions
+        // become candidates for sequencing.
+        let mut forward_index: BTreeMap<&TransactionDigest, Vec<&TransactionDigest>> =
+            BTreeMap::new();
+
+        // Keep track of the in-degree to facilitate topological sort.
+        let mut in_degree: HashMap<&TransactionDigest, usize> = HashMap::new();
+        let mut to_visit: BTreeSet<_> = effect_map.keys().collect();
+
+        for (d, effect) in &effect_map {
+            let entry = in_degree.entry(d).or_default();
+
+            for dep in &effect.dependencies {
+                // We only record the dependencies not in a checkpoint, as the ones
+                // in a checkpoint are already satisfied presumably.
+                if tx_not_in_checkpoint.contains(dep) {
+                    forward_index.entry(dep).or_default().push(d);
+
+                    // We record a dependency from within the tx not in a checkpoint
+                    *entry += 1;
+                }
+            }
+
+            // If it has a dependency it cannot be a starting item for the topological
+            // sort.
+            if *entry > 0 {
+                to_visit.remove(d);
+            }
+        }
+
+        // This implements the topological sort
+        // TODO: implement an order that allows for parallel execution,
```

Comment:
I don't think we have to worry about this - the sync code does things in parallel already. If things aren't ordered perfectly, it may be slightly slower but I don't think it will make a big difference. I'll be sending out a PR soon.